wts.go

```diff
@@ -0,0 +1,435 @@
+		Domain:                  syscall.UTF16ToString(i.Domain[:]),
+		UserName:                syscall.UTF16ToString(i.UserName[:]),
+		ConnectTime:             windowsFileTimeToTime(i.ConnectTime),
+		DisconnectTime:          windowsFileTimeToTime(i.DisconnectTime),
+		LastInputTime:           windowsFileTimeToTime(i.LastInputTime),
+		LogonTime:               windowsFileTimeToTime(i.LogonTime),
+		CurrentTime:             windowsFileTimeToTime(i.CurrentTime)}, nil
+}
+
+func (wts *WTS) QuerySessionAddressV4(sessionID uint32) (wrappers.WTS_CLIENT_ADDRESS, error) {
+	var buffer uintptr
+	var bytesReturned uint32
+
+	if err := wrappers.WTSQuerySessionInformation(wts.handle, sessionID, wrappers.WTSSessionAddressV4, &buffer, &bytesReturned); err != nil {
+		return wrappers.WTS_CLIENT_ADDRESS{}, err
+	}
+	defer wrappers.WTSFreeMemory(unsafe.Pointer(buffer))
+
+	if bytesReturned != 24 {
+		return wrappers.WTS_CLIENT_ADDRESS{}, buferSizeError(24, bytesReturned)
+	}
+	return *(*wrappers.WTS_CLIENT_ADDRESS)(unsafe.Pointer(buffer)), nil
+}
+
+func (wts *WTS) QuerySessionIsRemoteSession(sessionID uint32) (bool, error) {
+	return wts.querySessionInformationAsBool(sessionID, wrappers.WTSIsRemoteSession)
+}
+
+func (wts *WTS) QueryUserToken(sessionID uint32) (*Token, error) {
+	var handle syscall.Handle
+	if err := wrappers.WTSQueryUserToken(sessionID, &handle); err != nil {
+		return nil, err
+	}
+	return &Token{handle: handle}, nil
+}
+
+func clientAddressToIP(addressFamily uint32, address []byte) (net.IP, error) {
+	switch addressFamily {
+	case syscall.AF_INET:
+		if len(address) >= 4 {
+			return net.IPv4(address[0], address[1], address[2], address[3]), nil
+		}
+	case syscall.AF_INET6:
+		if len(address) >= 16 {
+			return net.IP(address[:16]), nil
+		}
+	}
+	return nil, fmt.Errorf("Unknown addressFamily: %v", addressFamily)
+}
+
+func buferSizeError(excpected, returned uint32) error {
+	return fmt.Errorf("Invalid buffer size. Expected: %d returned: %d", excpected, returned)
+}
+
+func utf16BufferToString(buffer *uintptr, len uint32) string {
+	return syscall.UTF16ToString((*[256]uint16)(unsafe.Pointer(*buffer))[:len])
+}
+
+func utf16PtrToString(s *uint16) string {
+	return syscall.UTF16ToString((*[256]uint16)(unsafe.Pointer(s))[:])
+}
+
+func windowsFileTimeToTime(fileTime int64) time.Time {
+	const TicksPerSecond = 10000000
+	const EpochDifference = 11644473600
+	// we also can use win32 api FileTimeToSystemTime
+	return time.Unix((fileTime/TicksPerSecond)-EpochDifference, 0)
+}
+
+func (wts *WTS) querySessionInformationAsBool(sessionID uint32, infoClass uint32) (bool, error) {
+	var buffer uintptr
+	var bytesReturned uint32
+
+	if err := wrappers.WTSQuerySessionInformation(wts.handle, sessionID, infoClass, &buffer, &bytesReturned); err != nil {
+		return false, err
+	}
+	defer wrappers.WTSFreeMemory(unsafe.Pointer(buffer))
+
+	if bytesReturned != 1 {
+		return false, buferSizeError(1, bytesReturned)
+	}
+	return *(*bool)(unsafe.Pointer(buffer)), nil
```

Comment:
I'm not sure if this is guaranteed to work.  It would probably be safer to retrieve the raw Win32 Boolean as an integer and do `!= 0` on it.